### PR TITLE
Bug: Atome failed due to sub product amount zero

### PR DESCRIPTION
**Pull request by ajzkk**
#### 1. Objective

Bug fixed Atome failed due to sub product amount zero

#### 2. Description of change

- Remove sub products with amount zero
- When sub total is zero, show error message `Complimentary products cannot be billed`.

#### 3. Quality assurance

Tested locally with Atome payment method when

- Sub total is zero
- Bundle product has sub product with amount zero
- And normal product order

#### 4. Impact of the change

N/A

#### 5. Priority of change

Normal

### CODE REVIEW

File: Gateway/Validator/APMRequestValidator.php [hunk 1]
```diff
@@ -81,6 +82,10 @@ private function validateAtomeAmount($order)
         $currency = strtoupper($order->getCurrencyCode());
         $amount = $order->getGrandTotalAmount();
 
+        if ($subTotal === 0.0) {
+            throw new LocalizedException(__('Complimentary products cannot be billed'));
```
Comment (danfowler) [inline Gateway/Validator/APMRequestValidator.php:86]: How do we decide what to translate and what not to translate?

File: Gateway/Request/APMBuilder.php [hunk 2]
```diff
@@ -375,16 +375,17 @@ private function getOrderItems($order)
...
-                'sku' => $item['sku'],
-                'name' => $item['name'],
-                'amount' => $this->money->setAmountAndCurrency(
-                    $item['base_original_price'],
-                    $currency
-                )->toSubunit(),
-                'quantity' => $item['qty_ordered'],
+                'sku' => $item->getSku(),
+                'name' => $item->getName(),
+                'amount' => $this->money->setAmountAndCurrency($price, $currency)->toSubunit(),
```
Comment (danfowler) [inline Gateway/Request/APMBuilder.php:387]: Did a linter trigger the formatting change?

File: Gateway/Request/APMBuilder.php [hunk 3]
```diff
@@ -375,16 +375,17 @@ private function getOrderItems($order)
         $itemArray = [];
         $items = $order->getItems();
         $currency = $order->getCurrencyCode();
-        foreach ($items as $itemObject) {
-            $item = $itemObject->toArray();
+
```
Comment (danfowler) [inline Gateway/Request/APMBuilder.php:378]: Did a linter trigger the formatting change?

File: Gateway/Request/APMBuilder.php (hunk 2 above)
Comment (ajzkk) [reply Gateway/Request/APMBuilder.php:387]: No. Currently we do not have linter in Magento.

Comment (ajzkk) [reply Gateway/Request/APMBuilder.php:387]: But we do have Magento 2 coding standard checker in our pipeline.

Comment (danfowler) [reply Gateway/Request/APMBuilder.php:387]: Oh, it would be nice to have a linter so our formatting changes are guaranteed to only go one way.

Comment (ajzkk) [reply Gateway/Request/APMBuilder.php:387]: The file is formatted with php-cs-fixer locally. 
Will add linter with separate ticket in order to share the same rule with team.

Comment (danfowler) [reply Gateway/Request/APMBuilder.php:387]: Great, thanks.

File: Gateway/Request/APMBuilder.php [hunk 4]
```diff
@@ -375,16 +375,17 @@ private function getOrderItems($order)
...
+        foreach ($items as $item) {
+            $price = $item->getPrice();
```
[the remark this answers is not in the document]
Comment (ajzkk) [reply Gateway/Request/APMBuilder.php:380]: `getPrice()` will return amount with it's sub items.

[the remark this answers is not in the document]
Comment (ajzkk) [reply Gateway/Request/APMBuilder.php:380]: previous we are using `base_original_price` which is only the price of parent item.

File: Gateway/Validator/APMRequestValidator.php [hunk 5]
```diff
@@ -81,6 +82,10 @@ private function validateAtomeAmount($order)
         $currency = strtoupper($order->getCurrencyCode());
         $amount = $order->getGrandTotalAmount();
 
+        if ($subTotal === 0.0) {
```
Comment (danfowler) [inline Gateway/Validator/APMRequestValidator.php:85]: Above, we cast `$price` to float. Is it not necessary here?

Comment (ajzkk) [reply Gateway/Validator/APMRequestValidator.php:85]: `getGrandTotalAmount` return float value by default and I have also tested too.

```
  /**
   * Returns order grand total amount
   *
   * @return float
   */
  public function getGrandTotalAmount();
```

Comment (ajzkk) [reply Gateway/Validator/APMRequestValidator.php:85]: let me cast to float for 100% sure.

**Review by danfowler — APPROVED** [review]

**Review by danfowler — APPROVED** [review]

**Review by aashishgurung — APPROVED** [review]

**Pull request merged**